# <center>Объявления о продаже автомобилей в Германии<center>

Вам представлен датасет с объявлениями о продаже в Германии на 2022 год. Необходимо осмотреть первично данные и выполнить их предобработку, также сделать соответствующие выводы по полученной информации

### Обзор данных

* `mileage` — Пройденное расстояние на автомобиле
* `make` — Марка автомобиля
* `model` — Модель автомобиля
* `fuel` — Тип топлива
* `gear` — Тип коробки передач
* `offerType` — Тип предложения, то есть «Новые», «Демонстрационные» машины и т.д.
* `price` — Цена продажи автомобиля в объявлении (в долларах)
* `hp` — Количество лошадиных сил
* `year` — Год производства автомобиля

Импортируйте необходимые библиотеки

In [2]:
import pandas as pd 

### Загрузка данных

1. Загрузите необходимый датасет в переменную

In [50]:
germany_cars=pd.read_csv('C:/Users/molot/YandexDisk/Колледж Настя/1Т/germany_cars.csv')
germany_cars

mileage        make   model               fuel    gear       offerType  \
0       235000         BMW     316             Diesel  Manual            Used   
1        92800  Volkswagen    Golf           Gasoline  Manual            Used   
2       149300        SEAT    Exeo           Gasoline  Manual            Used   
3        96200     Renault  Megane           Gasoline  Manual            Used   
4       156000     Peugeot     308           Gasoline  Manual            Used   
...        ...         ...     ...                ...     ...             ...   
46400       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46401       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46402       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46403       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46404       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   

         price     hp  year  
0       6800.0  116.0  2011  
1       6877.0  122.0  2011  
2       6900.0  160.0  2011  
3       6950.0  110.0  2011  
4       6950.0  156.0  2011  
...        ...    ...   ...  
46400  12990.0   71.0  2021  
46401  12990.0   71.0  2021  
46402  12990.0   71.0  2021  
46403  12990.0   71.0  2021  
46404  12990.0   71.0  2021  

[46405 rows x 9 columns]

2. Проведите первичный осмотр данных при помощи известных вам методов(дополнительно можно использовать метод hist() для оценки распределений). Сделайте соответствующие выводы на основе полученной информации

In [12]:
germany_cars.shape

(46405, 9)

In [13]:
germany_cars.sample(15)

mileage           make           model               fuel       gear  \
31465    30200           Ford            Kuga             Diesel  Automatic   
30671     3000           Audi             RS6           Gasoline  Automatic   
40708    48900  Mercedes-Benz        A 45 AMG           Gasoline  Automatic   
35105   185000        Citroen        Berlingo             Diesel     Manual   
24786    79252        Porsche             911           Gasoline  Automatic   
35121    48000     Volkswagen             up!           Gasoline     Manual   
16411    54232           Ford          Fiesta           Gasoline     Manual   
17807    89987     Volkswagen          Tiguan           Gasoline     Manual   
36845   177144     Volkswagen  Passat Variant             Diesel     Manual   
45357      200         Suzuki           Swift  Electric/Gasoline     Manual   
27700   147000        Peugeot             207           Gasoline     Manual   
44188     1000        Renault        Talisman             Diesel  Automatic   
35493   163973            BMW             525             Diesel  Automatic   
13129   145031         Nissan          Navara             Diesel     Manual   
32940    49986     Volkswagen  Golf Sportsvan           Gasoline  Automatic   

           offerType     price     hp  year  
31465           Used   23900.0  179.0  2018  
30671  Demonstration  157890.0  600.0  2021  
40708           Used   38450.0  381.0  2017  
35105           Used    5200.0   90.0  2012  
24786           Used   69700.0  349.0  2013  
35121           Used    5500.0   60.0  2012  
16411           Used    6990.0   60.0  2017  
17807           Used   10990.0  150.0  2011  
36845           Used   11499.0  140.0  2013  
45357           Used   14290.0   83.0  2021  
27700           Used    3000.0   95.0  2011  
44188  Demonstration   37180.0  200.0  2020  
35493           Used   12980.0  218.0  2012  
13129           Used   10686.0  190.0  2013  
32940           Used   17450.0  125.0  2017

In [11]:
germany_cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46405 entries, 0 to 46404
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   mileage    46405 non-null  int64  
 1   make       46405 non-null  object 
 2   model      46262 non-null  object 
 3   fuel       46158 non-null  object 
 4   gear       46223 non-null  object 
 5   offerType  46405 non-null  object 
 6   price      44872 non-null  float64
 7   hp         46376 non-null  float64
 8   year       46405 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 3.2+ MB


In [15]:
column_names=list(germany_cars.columns)
print("Названия колонок:", column_names)
print("Количество колонок:", len(column_names))

Названия колонок: ['mileage', 'make', 'model', 'fuel', 'gear', 'offerType', 'price', 'hp', 'year']
Количество колонок: 9


Выводы:
* В двтвфрейме 9 столбцов
* 5 из них имеют тип данных object
* 2 столбца имеют тип int64
* 2 столбца имеют тип float64
* Всего в датафрейме 46405 строк
* Объем занимаемой памяти 3.2+ MB

3. Переименуйте столбец `offerType` под стандартное название

In [19]:
new_column_names=[]
for word in column_names:
    if word!='offerType':
        new_column_names.append(word)
    else:
        new_column_names.append(word.lower())
germany_cars.columns=new_column_names
germany_cars.head()

mileage        make   model      fuel    gear offertype   price     hp  \
0   235000         BMW     316    Diesel  Manual      Used  6800.0  116.0   
1    92800  Volkswagen    Golf  Gasoline  Manual      Used  6877.0  122.0   
2   149300        SEAT    Exeo  Gasoline  Manual      Used  6900.0  160.0   
3    96200     Renault  Megane  Gasoline  Manual      Used  6950.0  110.0   
4   156000     Peugeot     308  Gasoline  Manual      Used  6950.0  156.0   

   year  
0  2011  
1  2011  
2  2011  
3  2011  
4  2011

### Работа с дубликатами данных

4. Необходимо вывести на экран количество дублированных строк в данных, если таковые имеются, а также избавиться от них. Не забудьте сбросить индекс в датафрейме при необходимости.

In [34]:
lenght=0
for g in germany_cars.duplicated():
    if g==True:
        lenght+=1
print("количество дубликатов в датафрейме:",lenght) #2003
germany_cars.drop_duplicates() #Удаление дубликатов
germany_cars

количество дубликатов в датафрейме: 2003


mileage        make   model               fuel    gear       offerType  \
0       235000         BMW     316             Diesel  Manual            Used   
1        92800  Volkswagen    Golf           Gasoline  Manual            Used   
2       149300        SEAT    Exeo           Gasoline  Manual            Used   
3        96200     Renault  Megane           Gasoline  Manual            Used   
4       156000     Peugeot     308           Gasoline  Manual            Used   
...        ...         ...     ...                ...     ...             ...   
46400       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46401       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46402       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46403       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   
46404       99        Fiat     500  Electric/Gasoline  Manual  Pre-registered   

         price     hp  year  
0       6800.0  116.0  2011  
1       6877.0  122.0  2011  
2       6900.0  160.0  2011  
3       6950.0  110.0  2011  
4       6950.0  156.0  2011  
...        ...    ...   ...  
46400  12990.0   71.0  2021  
46401  12990.0   71.0  2021  
46402  12990.0   71.0  2021  
46403  12990.0   71.0  2021  
46404  12990.0   71.0  2021  

[46405 rows x 9 columns]

### Работа с пропусками

5. Выведите на экран количество пропусков в каждом столбце, опишите их, почему они могут возникнуть и как их возможно исправить

In [41]:
germany_cars.isnull().sum()

mileage         0
make            0
model         143
fuel          247
gear          182
offerType       0
price        1533
hp             29
year            0
dtype: int64

В 5 столбцах есть пропуски. Они могли вознкнуть в связи с отсутствитем данных изначально. Чтобы исправить ситуацию нужжно проанализировать какие данные пропущены и выбрат способ заполнения, например заполнить пропуски средними или медианными значениями, заполнить словом "неизвестно" или другим подобным словом, заполнить отрацательными значениями (Yes, No)


6. Заполните пропуски в `model`,`fuel`,`gear`

In [61]:
germany_cars.model.fillna('Unknown', inplace=True)
germany_cars.fuel.fillna('Unknown', inplace=True)
germany_cars.gear.fillna('Unknown', inplace=True)

7. Удалите оставшиеся пропуски в столбцах `hp` и `price`.

In [149]:

germany_cars.hp.fillna('', inplace=True)
germany_cars.price.fillna('', inplace=True)

Проверьте, все ли пропуски точно удалились из датафрейма

In [150]:
germany_cars.isna().sum()

mileage       0
make          0
model         0
fuel          0
gear          0
offerType     0
price         0
hp            0
year          0
make_model    0
dtype: int64

8. Оцените количество уникальных марок автомобилей (производителей) с помощью метода unique() и выведите их количество числом (пожалуйста, не считайте вручную)

In [141]:
germany_cars_unicue=len(germany_cars.make.unique())

77

9. Выведите топ 20 марок автомобилей (производителей) по количеству объявлений с помощью метода value_counts()

In [151]:
germany_cars.model.value_counts().head(20)

model
Golf       1492
Corsa      1491
Fiesta     1289
Astra      1191
Focus       985
up!         945
Fabia       917
Polo        893
forTwo      725
500         666
Octavia     665
Clio        658
Ibiza       650
Twingo      646
Leon        574
Tiguan      570
i10         515
Aygo        499
A4          488
A3          482
Name: count, dtype: int64

### Группировка данных

10. Сгруппируйте данные по столбцу `make`, выведите для каждого производителя среднюю стоимость их автомобиля, а также среднее число лошадиных сил и отсортируйте по убыванию столбца `price`. Отобразите только топ 20 самых дорогих производителей. Напишите вывод, всё ли в порядке с полученными данными или Вас что-то смущает

In [128]:
germany_cars.groupby(by='make', as_index=False)[['price','hp']].agg("mean").sort_values(by='price', ascending=False).reset_index().head(20)

index         make          price          hp
0      46      Maybach  450479.333333  543.000000
1      27      Ferrari  324028.181818  693.909091
2      39  Lamborghini  305698.777778  690.111111
3      48      McLaren  212474.000000  646.636364
4      62  Rolls-Royce  194900.000000  605.000000
5      10      Bentley  192123.307692  554.866667
6       6        Aston  160063.166667  531.833333
7      17     Corvette  110160.000000  687.000000
8      59      Porsche   89477.470588  410.844262
9      45     Maserati   72480.583333  383.333333
10     26       FISKER   69900.000000  408.000000
11      4       Alpina   66046.300000  443.000000
12     58     Polestar   64150.000000  408.000000
13     69        Tesla   63163.782609  494.416667
14     52       Morgan   61950.000000  177.000000
15      5       Alpine   53783.000000  244.800000
16     41         Land   51338.018987  250.591463
17     60          RAM   47945.000000  401.000000
18     12     Cadillac   45032.142857  348.285714
19     24        Dodge   44968.818182  395.869565

11. Создайте новый столбец из `make` и `model` и выведите топ 10 самый дорогих и дешевых автомобилей с помощью средней стоимости

In [140]:
germany_cars['make_model']=germany_cars.make+" "+germany_cars.model
max=germany_cars.groupby(by='make_model', as_index=False,)['price'].agg('mean').sort_values('price', ascending=False).head(10).reset_index()
pd.options.display.float_format='{:.2f}'.format
max

index             make_model      price
0    237            Ferrari F12 1199900.00
1    572    Mercedes-Benz S 650  717078.00
2    448        Maybach Pullman  608439.00
3    236            Ferrari 812  439900.00
4     26           Aston Martin  398000.00
5    370  Lamborghini Aventador  371423.17
6    462          McLaren 675LT  297900.00
7    238     Ferrari F8 Tributo  294416.67
8    150    Bentley Flying Spur  251560.00
9    463           McLaren 720S  251145.00

In [135]:
germany_cars['make_model']=germany_cars.make+" "+germany_cars.model
germany_cars.groupby(by='make_model', as_index=False)['price'].agg('mean').sort_values('price').head(10).reset_index()

index                make_model        price
0    366               Lada Priora  1190.000000
1    153            Brilliance BS4  1500.000000
2    729                SEAT Panda  1650.000000
3    650           Others Landwind  1850.000000
4    865             Volvo Unknown  2499.000000
5    305              Hyundai Getz  2800.000000
6    800  Trailer-Anhänger Unknown  2800.000000
7    811            Volkswagen Fox  2920.230769
8    655               Peugeot 206  3393.934783
9    253                Fiat Linea  3400.000000

12. Сделайте общий вывод о проделанной работе

Датасет имеет хорошее качество данных после обработки, несмотря на наличие начальных пропусков и дубликатов. Рынок поддержаных автомобилей значитель преобладает над новым, что характерно для автомобильного рынка в целом. Цена автомобиля зависит не только от мощности, но и от других факторов, включая бренд, состояние, год выпуска и так далее.
ПРи анализе дата сета выявлен самый дорогой автомобиль и это Ferrari F12, самый дешевый автомобиль - Lada Priora. Самой популярной моделью является Golf, наименее популярной модель A3. В процессе анализа было заполнено 143 пропуска в столбцу с моделями, 247 пропусков в столбце с топливом и 182 просуека в столбце с названием механизма.
Анализ датасета может быть полезен для дальнейших маркетинговых исследований, оценки рыночных тенденций и ценообразования на автомобильном рынке Германии.